### PR TITLE
#11207: Introduce v1::DeviceHandle

### PR DESCRIPTION
**Pull request by patrickroberts**
### Ticket
#11207

### Problem description
In the v1 tt_metal API, objects will be replaced with opaque handles. To support the migration phase, opaque handles currently support conversion to object pointers, but that conversion will be removed along with v0 namespace.

### What's changed
Refactor API of DevicePool to return `v1::DeviceHandle` instead of `v0::Device *`

### Checklist
- [X] Post commit CI passes (https://github.com/tenstorrent/tt-metal/actions/runs/11406767460)
- [ ] Blackhole Post commit (if applicable)
- [X] Model regression CI testing passes
  - [(Single-card) Model perf tests](https://github.com/tenstorrent/tt-metal/actions/runs/11407183869)
  - [(T3K) T3000 model perf tests](https://github.com/tenstorrent/tt-metal/actions/runs/11407109435)
- [ ] Device performance regression CI testing passes (if applicable)
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

**Comment by cfjchu** [comment]
Neat trick. Overall looks good.

**Review by cfjchu — APPROVED** [review]

File: tt_metal/impl/device/device_pool.cpp
Comment (patrickroberts) [reply tt_metal/impl/device/device_pool.cpp:348]: Fair question. Once we remove the `v0` API, I'll be able to swap out the vector for a SlotMap, but in order for v0 and v1 to be compatible during this migration period, SlotMap (rightfully) prevents me from using just an index to refer to a Device.

**Review by pgkeller — COMMENTED** [review]
this looks fine to me.  let's hold off a bit longer before merging, need to be sure all ducks are in a row...

**Review by aliuTT — APPROVED** [review]

**Pull request merged**
